Optim foreach cleanup for SGD

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* __->__ #70481
* #70295

Add foreach flag to SGD optimizer + cleanup

Differential Revision: [D33767868](https://our.internmc.facebook.com/intern/diff/D33767868)

mikaylagawarecki committed:
Optim foreach cleanup for SGD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

albanD commented on torch/optim/sgd.py:
```diff
@@ -123,37 +128,164 @@ def step(self, closure=None):
+def sgd(params: List[Tensor],
+        d_p_list: List[Tensor],
+        momentum_buffer_list: List[Optional[Tensor]],
+        has_sparse_grad: bool = False,
```
Same question about has sparse grad default value

albanD reviewed (commented):
Sounds good. May need update depending on the has sparse grad discussion on the other PR.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/sgd.py:
```diff
@@ -123,37 +128,174 @@ def step(self, closure=None):
+                has_sparse_grad=has_sparse_grad,
+                foreach=group['foreach'])
 
             # update momentum_buffers in state
```
This is necessary here because we pass None when initializing momentum_buffer, so the update in the functional does not alias. I didn't fold this into the functional and left it like this.

Is there a good way of passing a tensor of the shape of grad that signifies `None`? (would something zeros_like(p.grad) and then a zero check in the functional potentially cause issues?)

albanD replied:
I think we can significantly simplify all this by doing the initialization here (by initializing with `dampening / momentum * grad` and removing all the special case in the functional variant). But yes that can be left as future work for now.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"




[ghstack-poisoned]

albanD reviewed (approved):
:+1:

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767868).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"


Differential Revision: [D33767868](https://our.internmc.facebook.com/intern/diff/D33767868)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767868).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for SGD"


Differential Revision: [D33767868](https://our.internmc.facebook.com/intern/diff/D33767868)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767868).
